source/_posts/2015-08-18-elasticsearch-at-househappy.markdown

```diff
@@ -0,0 +1,162 @@
+
+See the entire [JSON query here](https://gist.github.com/moxley/d8935387133476db0ba0).
+It's 174 lines of formatted JSON. Elasticsearch's query DSL is not the
+prettiest language I've seen. It kind of reminds me of [Postscript](https://en.wikipedia.org/wiki/PostScript), in that it is a language
+meant to be written by programs instead of programmers.
+
+### "Gridview" Search Results
+
+![Primary Search Results](http://cl.ly/image/1M2M1a2s3Y1L/search-results.jpg)
+
+The primary ("gridview") search generates a much more simple Elasticsearch
+query than the autocomplete. It finds properties with an exact-match by foreign key:
+
+```json
+{
+  "query": {
+    "bool": {
+      "must": [
+        {
+          "in": {
+            "city_id": "32"
+          }
+        },
+        {
+          "in": {
+            "public_status": "available"
+          }
+        }
+      ]
+    }
+  }
+}
+```
+
+In the above example, it searches Property Listings belonging to `city_id=32` (Portland, OR),
+with a public status of `available`. We also add filters and sorting to this
+query, based on the filters and sorts the user selects.
+
+Before using Elasticsearch for the primary search, we queried PostgreSQL
+directly. The primary table has millions of records, and the search query joined as
+many as 10 other tables. After switching to Elasticsearch for the query, using
+denormalized data, site performance improved dramatically.
+
+![Performance Before and After](https://pbs.twimg.com/media/CMDxrGWVEAA4k7a.jpg:large)
+
+## How Documents are Added
+
+### Converting Models to Documents
+
+We use the amazing
+[representable gem](https://github.com/apotonick/representable) to
+convert an ActiveRecord instance to an Elasticsearch document. This is the
+same gem we use to serialize ActiveRecord instances to API responses, and to
+deserialize API requests to ActiveRecord instances.
+
+We define a "representer" that essentially defines the schema of the
+document. That is, it describes which model attributes and which associations,
+and which association attributes will be in the document. It can also add
+virtual attributes that don't even exist on the model.
+Once the representer is defined, it's a simple piece of code to convert a model
+instance the Elasticsearch document JSON.
+
+```ruby
+metro_area = MetroArea.first
+representer = MetroAreaRepresenter.new(metro_area)
+metro_area_json = representer.to_json
+```
+
+The resulting JSON:
+
+```json
+{
+  "id": 338,
+  "name": "Portland-Vancouver-Hillsboro, OR-WA",
+  "location": "POINT (-122.4843457 45.6006204)",
+  "viewport": "..."
+}
+```
+
+We use the [elasticsearch gem](https://github.com/elastic/elasticsearch-ruby)
+to save the document to Elasticsearch.
+
+### Batch Loading
+
+For every type of Elasticsearch document we store, we created a `rake` task
+that loads every record of that type from PostgreSQL to
+Elasticsearch.
+
+### Incremental Synchronizing
```

Comment:
this bit could benefit from some flowchart or something to better demo how these workers communicate